File: app/src/main/java/io/github/datwheat/asthtc/MainActivity.java
package io.github.datwheat.asthtc;

import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v4.app.NotificationCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.PopupMenu;
import android.util.Log;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.ibm.icu.text.Transliterator;
import com.jakewharton.rxbinding2.view.RxView;
import com.jakewharton.rxbinding2.widget.RxTextView;
import com.jakewharton.rxbinding2.widget.TextViewTextChangeEvent;

import java.util.Random;

import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;


public class MainActivity extends AppCompatActivity {
    public static final String NOTI_ID = "NOTI_ID";
    private final String TAG = MainActivity.class.getName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final Transliterator t = Transliterator.getInstance("Halfwidth-Fullwidth");

        EditText editText = ((EditText) findViewById(R.id.boringEditText));
        final TextView betterTextView = (TextView) findViewById(R.id.betterTextView);
        ImageButton moreButton = (ImageButton) findViewById(R.id.imageButton);
        Button copyToClipboardButton = (Button) findViewById(R.id.button);


        RxTextView.textChangeEvents(editText)
                .subscribe(new Observer<TextViewTextChangeEvent>() {
                    @Override
                    public void onSubscribe(Disposable d) {
                        Log.i(TAG, "onSubscribe: Yee weouchea");
                    }

                    @Override
                    public void onNext(TextViewTextChangeEvent value) {
                        Log.i(TAG, "onNext: " + value.text().toString());
                        String betterText = t.transliterate(value.text().toString());
                        betterTextView.setText(betterText);
                    }

                    @Override
                    public void onError(Throwable e) {
                        Log.e(TAG, "onError: " + e.toString());
                    }

                    @Override
                    public void onComplete() {
                        Log.i(TAG, "onComplete: Yee wedone");
                    }
                });

        RxView.clicks(copyToClipboardButton).subscribe(new Observer<Object>() {
            @Override
            public void onSubscribe(Disposable d) {
                Log.i(TAG, "onSubscribe: Yee weouchea");
            }

            @Override
            public void onNext(Object obj) {
                String betterText = betterTextView.getText().toString();

                if (betterText.length() > 0) {
                    ClipboardManager clipboard = (ClipboardManager)
                            getSystemService(Context.CLIPBOARD_SERVICE);

                    ClipData clip = ClipData.newPlainText("asthtc txt", betterText);

                    clipboard.setPrimaryClip(clip);

                    Toast.makeText(MainActivity.this, "Text Copied", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onError(Throwable e) {
                Log.e(TAG, "onError: " + e.toString());
            }

            @Override
            public void onComplete() {
                Log.i(TAG, "onComplete: Yee wedone");
            }
        });

        moreButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                PopupMenu popup = new PopupMenu(MainActivity.this, view);

                MenuInflater inflater = popup.getMenuInflater();
                inflater.inflate(R.menu.main_activity_actions, popup.getMenu());

                popup.show();

                popup.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                    @Override
                    public boolean onMenuItemClick(MenuItem item) {
                        switch (item.getItemId()) {
                            case R.id.new_notification:
                                return setNotification();
                            case R.id.remove_notification:
                                return removeNotification();
                            default:
                                return false;
                        }
                    }
                });
            }
        });

    }

    private boolean setNotification() {
        removeNotification();
        SharedPreferences settings = getPreferences(MODE_PRIVATE);
        int newNotiId = (new Random()).nextInt();

        NotificationCompat.Builder builder =
                new NotificationCompat.Builder(this)
                        .setSmallIcon(R.drawable.ic_stat_panorama_fish_eye)
                        .setContentTitle(getString(R.string.app_name))
                        .setOngoing(true)
                        .setContentText("Tap to generate ａｅｓｔｈｅｔｉｃ text!");

        Intent resultIntent = new Intent(this, MainActivity.class);

        PendingIntent resultPendingIntent =
                PendingIntent.getActivity(
                        this,
                        0,
                        resultIntent,
                        PendingIntent.FLAG_UPDATE_CURRENT
                );

        builder.setContentIntent(resultPendingIntent);

        NotificationManager mNotifyMgr = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);

        mNotifyMgr.notify(newNotiId, builder.build());

        settings.edit().putInt(NOTI_ID, newNotiId).apply();
        return true;
    }

    private boolean removeNotification() {
        int notiId = getNoticationId();
        if (notiId != 0) {
            NotificationManager notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
            notificationManager.cancel(notiId);

            SharedPreferences settings = getPreferences(MODE_PRIVATE);
            settings.edit().putInt(NOTI_ID, 0).apply();
        }
        return true;
    }

    private int getNoticationId() {
        SharedPreferences settings = getPreferences(MODE_PRIVATE);
        return settings.getInt(NOTI_ID, 0);
    }

}

File: app/src/main/java/io/github/datwheat/asthtc/MainActivity.kt
package io.github.datwheat.asthtc

import android.app.NotificationManager
import android.app.PendingIntent
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v4.app.NotificationCompat
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.PopupMenu
import android.util.Log
import android.widget.Toast
import com.ibm.icu.text.Transliterator
import com.jakewharton.rxbinding2.view.RxView
import com.jakewharton.rxbinding2.widget.RxTextView
import com.jakewharton.rxbinding2.widget.TextViewTextChangeEvent
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import kotlinx.android.synthetic.main.activity_main.*
import java.util.*


class MainActivity : AppCompatActivity() {
    private val TAG = MainActivity::class.java.name

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        val t = Transliterator.getInstance("Halfwidth-Fullwidth")

        RxTextView.textChangeEvents(boringEditText)
                .subscribe(object : Observer<TextViewTextChangeEvent> {
                    override fun onSubscribe(d: Disposable) {
                        Log.i(TAG, "onSubscribe: Yee weouchea")
                    }

                    override fun onNext(value: TextViewTextChangeEvent) {
                        Log.i(TAG, "onNext: " + value.text().toString())
                        val betterText = t.transliterate(value.text().toString())
                        betterTextView.text = betterText
                    }

                    override fun onError(e: Throwable) {
                        Log.e(TAG, "onError: " + e.toString())
                    }

                    override fun onComplete() {
                        Log.i(TAG, "onComplete: Yee wedone")
                    }
                })

        RxView.clicks(copyToClipboardButton).subscribe(object : Observer<Any> {
            override fun onSubscribe(d: Disposable) {
                Log.i(TAG, "onSubscribe: Yee weouchea")
            }

            override fun onNext(obj: Any) {
                val betterText = betterTextView.text.toString()

                if (betterText.isNotEmpty()) {
                    val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager

                    clipboard.primaryClip = ClipData.newPlainText("asthtc txt", betterText)

                    Toast.makeText(this@MainActivity, "Text Copied", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onError(e: Throwable) {
                Log.e(TAG, "onError: " + e.toString())
            }

            override fun onComplete() {
                Log.i(TAG, "onComplete: Yee wedone")
            }
        })

        moreButton.setOnClickListener { view ->
            val popup = PopupMenu(this@MainActivity, view)

            val inflater = popup.menuInflater
            inflater.inflate(R.menu.main_activity_actions, popup.menu)

            popup.show()

            popup.setOnMenuItemClickListener { item ->
                when (item.itemId) {
                    R.id.new_notification -> setNotification()
                    R.id.remove_notification -> removeNotification()
                    else -> false
                }
            }
        }
    }

    private fun setNotification(): Boolean {
        removeNotification()
        val settings = getPreferences(Context.MODE_PRIVATE)
        val newNotiId = Random().nextInt()

        val builder = NotificationCompat.Builder(this)
                .setSmallIcon(R.drawable.ic_stat_panorama_fish_eye)
                .setContentTitle(getString(R.string.app_name))
                .setOngoing(true)
                .setColor(ContextCompat.getColor(this, R.color.majorelle_blue))
                .setContentText("Tap to generate ａｅｓｔｈｅｔｉｃ text!")

        val resultIntent = Intent(this, MainActivity::class.java)

        val resultPendingIntent = PendingIntent.getActivity(
                this,
                0,
                resultIntent,
                PendingIntent.FLAG_UPDATE_CURRENT
        )

        builder.setContentIntent(resultPendingIntent)

        val mNotifyMgr = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager

        mNotifyMgr.notify(newNotiId, builder.build())

        settings.edit().putInt(NOTI_ID, newNotiId).apply()
        return true
    }

    private fun removeNotification(): Boolean {
        val notiId = noticationId
        if (notiId != 0) {
            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notificationManager.cancel(notiId)

            val settings = getPreferences(Context.MODE_PRIVATE)
            settings.edit().putInt(NOTI_ID, 0).apply()
        }
        return true
    }

    private val noticationId: Int
        get() {
            val settings = getPreferences(Context.MODE_PRIVATE)
            return settings.getInt(NOTI_ID, 0)
        }

    companion object {
        val NOTI_ID = "NOTI_ID"
    }

}
